daemon/container_operations.go

```diff
@@ -957,68 +957,101 @@ func (daemon *Daemon) ConnectToNetwork(container *container.Container, idOrName
-	} else {
-		if err := daemon.connectToNetwork(container, idOrName, endpointConfig, true); err != nil {
+
+		if err := daemon.updateNetworkConfig(container, n, endpointConfig, true); err != nil {
 			return err
 		}
+	} else if !daemon.isNetworkHotPluggable() {
+		return fmt.Errorf(runtime.GOOS + " does not support connecting a running container to a network")
+	} else if err := daemon.connectToNetwork(container, idOrName, endpointConfig, true); err != nil {
+		return err
 	}
+
 	if err := container.ToDiskLocking(); err != nil {
 		return fmt.Errorf("Error saving container to disk: %v", err)
 	}
 	return nil
 }
 
-// DisconnectFromNetwork disconnects container from network n.
+// DisconnectFromNetwork disconnects container from network.
 func (daemon *Daemon) DisconnectFromNetwork(container *container.Container, networkName string, force bool) error {
-	n, err := daemon.FindNetwork(networkName)
-	if !container.Running || (err != nil && force) {
-		if container.RemovalInProgress || container.Dead {
-			return errRemovalContainer(container.ID)
-		}
-		// In case networkName is resolved we will use n.Name()
-		// this will cover the case where network id is passed.
-		if n != nil {
-			networkName = n.Name()
-		}
-		if _, ok := container.NetworkSettings.Networks[networkName]; !ok {
-			return fmt.Errorf("container %s is not connected to the network %s", container.ID, networkName)
-		}
-		delete(container.NetworkSettings.Networks, networkName)
-	} else if err == nil && !daemon.isNetworkHotPluggable() {
-		return fmt.Errorf(runtime.GOOS + " does not support connecting a running container to a network")
-	} else if err == nil {
-		if container.HostConfig.NetworkMode.IsHost() && containertypes.NetworkMode(n.Type()).IsHost() {
-			return runconfig.ErrConflictHostNetwork
-		}
+	if !container.Running {
+		return daemon.DisconnectStoppedContainerFromNetwork(container, networkName)
+	}
+	return daemon.DisconnectRunningContainerFromNetwork(container, networkName, force)
+}
+
+// DisconnectStoppedContainerFromNetwork disconnects a stopped container from network.
+func (daemon *Daemon) DisconnectStoppedContainerFromNetwork(container *container.Container, networkName string) error {
+	if container.RemovalInProgress || container.Dead {
+		return errRemovalContainer(container.ID)
+	}
 
-		if err := daemon.disconnectFromNetwork(container, n, false); err != nil {
+	network, err := daemon.FindNetwork(networkName)
+	if err != nil {
+		// when no such network and container is stopped,
+		// we should consider such a situation that
+		// this network is removed while network details may be still
+		// in networking config of a stopped container.
+		if _, exist := container.NetworkSettings.Networks[networkName]; !exist {
+			// if stopped container has no this non-found network details, return no such network error
 			return err
 		}
+		delete(container.NetworkSettings.Networks, networkName)
```

Comment:
Have only one `delete(container.NetworkSettings.Networks, networkName)` after the if block.